### PR TITLE
New module: Jenkins node management

### PR DESCRIPTION
**Pull request by phyrwork**
##### SUMMARY
Add `jenkins_node` module for managing Jenkins nodes via `python-jenkins` REST-API package a la the other `jenkins_...` modules. 

Implement basic states of `present` (create), `absent` (delete), `disabled` (offline), `enabled` (online).

Implement idempotent configuration of

- Almost fully, top-level node options including environment variables and tool locations
- Almost fully, SSH launcher options
- Some JNLP launcher options

After node creation all configuration is done by updating config.xml by the REST API.

All configuration is opt-in - only options specified will be changed on the Jenkins node.

Best effort taken to only issue a configuration change if we believe we have changed a value.

Where options configure plugins e.g. choice of launcher, API is designed to be extensible. See examples - SSH launcher and SSH host key validation. **Some advice on how to document parameters beyond the top-level would be appreciated. Best I could figure so far was to show everything in examples.**

**Tests pending, but all working as intended against local Jenkins instance  - submitting for some initial feedback before I get too far into the tests.**

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

jenkins_node

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by felixfontein** [comment]
Thanks for your contribution! In case you haven't, please check out the section on adding new modules in CONTRIBUTING.md: https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins

> Some advice on how to document parameters beyond the top-level would be appreciated.

Grep for "suboptions:". You can use it when `type=dict` or `type=list` and `elements=dict`, for example like this: https://github.com/ansible-collections/community.general/blob/main/plugins/modules/nmcli.py#L129-L164

**Comment by phyrwork** [comment]
> Grep for "suboptions:". You can use it when `type=dict` or `type=list` and `elements=dict`, for example like this: https://github.com/ansible-collections/community.general/blob/main/plugins/modules/nmcli.py#L129-L164

Thanks, that will get me some of the way there.

However some of my options e.g. `launcher` select and configure Jenkins plugin `type`s with different sets of `args`.

e.g. from the docs (and then expanded a bit to make the point)

```
- name: Use known hosts file for host key verification when connecting to a node by SSH
  community.general.jenkins_node:
    name: my-agent
    launcher:
      type: ssh
      args:
        host: my-agent.test
        port: 22
        credentials_id: 0bf9599f-b551-4afe-abae-cde7b14a406c
        host_key_verify:
          type: content
          args:
            algorithm: ssh-rsa
            key: AAAAB3NzaC1yc2EAAAADAQABAAABgQC...

- name: Set Jenkins node to launch by connecting it to the controller
  community.general.jenkins_node:
    name: my-agent
    launcher:
      type: jnlp
      args:
        data_dir: remoting
        work_dir: custom
        disable_work_dir: yes
        assert_work_dir: no
        use_web_socket: no
```

Internally there are separate `ArgumentSpecValidator`s for each `args` set, so that at the top-level the `args` type can just be `dict`, but of course the types need enumerating and their options documented in full somewhere.

Can you suggest how I could do this?

**Comment by felixfontein** [comment]
> However some of my options e.g. `launcher` select and configure Jenkins plugin `type`s with different sets of `args`.
> 
> e.g. from the docs (and then expanded a bit to make the point)

There is no way to properly document this, except by using examples and annotating them with comments.

What one usually does is provide one sub-option for every type, like `args_ssh`, `args_jnlp`. Then you can require `args_ssh` if `type=ssh`, require `args_jnlp` if `tpye=jnlp`, and make `args_ssh` and `args_jnlp` mutually exclusive. See for example https://docs.ansible.com/ansible/latest/collections/community/hrobot/boot_module.html#parameters with the various `install_*` options (there, type is implicitly set by "which `install_*` option is set?").

File: plugins/modules/jenkins_node.py [hunk 1]
```diff
@@ -0,0 +1,1311 @@
...
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: jenkins_node
+short_description: Manage Jenkins nodes
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:14]:
```suggestion
short_description: Manage Jenkins nodes
version_added: 7.4.0
```

File: plugins/modules/jenkins_node.py [hunk 2]
```diff
@@ -0,0 +1,1311 @@
...
+      - Password to authenticate with the Jenkins server.
+    type: str
+  token:
+    description:
+      - API token to authenticate with the Jenkins server.
+    type: str
+  state:
+    description:
+      - Specifies whether the Jenkins node should be V(present) (created), V(absent)
+        (deleted), enabled (online) or disabled (offline).
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:54]:
```suggestion
        (deleted), V(enabled) (online) or V(disabled) (offline).
```

File: plugins/modules/jenkins_node.py [hunk 3]
```diff
@@ -0,0 +1,1311 @@
...
+  labels:
+    description:
+      - When specified, sets the Jenkins node labels.
+    type: list
+    elements: str
+  mode:
+    description:
+      - When specified, sets the Jenkins node usage policy.
+      - V(normal) means use node as much as possible.
+      - V(exclusive) means only build jobs with label expressions that match tnode.
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:83]: Should `tnode` be `the node` or `node`? Or is `tnode` some Jenkins specific term?

File: plugins/modules/jenkins_node.py [hunk 4]
```diff
@@ -0,0 +1,1311 @@
...
+    type: str
+    choices:
+      - normal
+      - exclusive
+  launcher:
+    description:
+      - When specified, sets the Jenkins node launch method.
+      - Launcher V(type) must be specified.
+      - V(args) are specified according to the launcher V(type) (see examples).
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:92]:
```suggestion
      - Launcher O(launcher.type) must be specified.
      - V(args) are specified according to the launcher O(launcher.type) (see examples).
```

File: plugins/modules/jenkins_node.py [hunk 5]
```diff
@@ -0,0 +1,1311 @@
...
+    type: dict
+    suboptions:
+      type:
+        description:
+          - Specifies the launch method.
+        type: str
+        choices:
+          - ssh
+          - jnlp
+        required: yes
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:102]: Always use `true` and `false for YAML booleans.
```suggestion
        required: true
```
There are multiple more in this file.

File: plugins/modules/jenkins_node.py [hunk 6]
```diff
@@ -0,0 +1,1311 @@
...
+    type: dict
+  tools:
+    description:
+      - When specified, sets the Jenkins node tool locations.
+    type: list
+    elements: dict
+    suboptions:
+      type:
+        description:
+          - Jenkins tool class name e.g. "hudson.plugins.git.GitTool$DescriptorImpl" for
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:126]:
```suggestion
          - Jenkins tool class name, for example V(hudson.plugins.git.GitTool$DescriptorImpl) for
```

File: plugins/modules/jenkins_node.py [hunk 7]
```diff
@@ -0,0 +1,1311 @@
...
+         type: str
+         required: yes
+      home:
+        description:
+          - Jenkins tool home directory.
+        type: path
+        required: yes
+  offline_message:
+    description:
+      - Specifies the offline reason message to be set when setting a Jenkins node state
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:142]:
```suggestion
      - Specifies the offline reason message to be set when setting a Jenkins node O(state)
```

File: plugins/modules/jenkins_node.py [hunk 8]
```diff
@@ -0,0 +1,1311 @@
...
+    - Manage Jenkins nodes with Jenkins REST API.
+requirements:
+  - "python-jenkins >= 0.4.12"
+author:
+  - Connor Newton (@phyrwork)
+extends_documentation_fragment:
+  - community.general.attributes
+attributes:
+  check_mode:
+    support: none
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:25]: What is the reason to not implement check mode?

**Comment by phyrwork** [comment]
Gradually chipping away at the unit tests when I have time...

File: plugins/modules/jenkins_node.py [hunk 9]
```diff
@@ -0,0 +1,1383 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: jenkins_node
+short_description: Manage Jenkins nodes
+version_added: 7.4.0
```
Comment (felixfontein) [inline plugins/modules/jenkins_node.py:15]:
```suggestion
version_added: 8.0.0
```

**Comment by felixfontein** [comment]
Ping @phyrwork 

needs_info

**Comment by phyrwork** [comment]
Hey. I haven't been able to work on this for some time now and am unlikely to again for a while.

What is missing is unit test coverage for several of the fields implemented - these are tedious to implement and have taken a long time.

I could cut these fields from the PR to be finished later and tidy up what is done... unless it would be acceptable to submit with less than complete coverage.

**Comment by felixfontein** [comment]
The tests don't need to cover *everything*, they look extensive enough to me to warrant a merge.

But you definitely need to fix the failing tests.

**Comment by felixfontein** [comment]
@phyrwork ping

needs_info

**Pull request closed without merging**